### PR TITLE
range{feed,client},roachpb: standardize mock generation through bazel

### LINKED ISSUE #67010: bazel: consider better patterns for mock generation
In #66910 I added a bunch of interface mocks using gomock/mockgen. The Bazel build integration got a bit messy because of dependency handling, we should revisit this and come up with a best practice pattern for generating mocks.

/cc @cockroachdb/dev-inf @irfansharif

### PR DESCRIPTION
**Pull request by irfansharif**
Fixes #67010. Also clean up some rotted gomock generation in sqlproxyccl
(mockgen usage was removed in #66369).

Release note: None

### CODE REVIEW

**Comment by irfansharif** [comment]
Actually, hold off on looking. Clearly broke something and only just figured out how to run the bazel docker image locally.

**Comment by irfansharif** [comment]
I'm two ways about this PR. The benefit it offers over the status quo (currently checked into master) is that gazelle will now automatically keep the "base" go_library targets up-to-date. For e.g., right now any file/dep changes to `package rangefeed` require manual edits to the package's BUILD.bazel, to keep `[rangefeed_base](https://github.com/cockroachdb/cockroach/blob/3fce3274a43864f412fd7c5933790af50f18938e/pkg/kv/kvclient/rangefeed/BUILD.bazel#L4-L28)` up-to-date. 

Though it's a bit unfortunate the kind of funky wiring we needed here to make that happen. I feel it's mostly working around upstream "bugs" (https://github.com/bazelbuild/bazel-gazelle/issues/1077, https://github.com/bazelbuild/bazel-gazelle/issues/1078), so perhaps it's acceptable?

File: pkg/kv/kvclient/kvcoord/BUILD.bazel
Comment (rickystewart) [inline pkg/kv/kvclient/kvcoord/BUILD.bazel:99]: This is pretty ugly. I think you're having to do this because `pkg/kv/kvclient/kvcoord` is a package that contains both [internal and external tests](https://github.com/bazelbuild/rules_go/blob/master/go/core.rst#internal-test-example), right?

The blessed way to do that would be to have two separate test targets, one for the external tests and one for the internal tests. Unfortunately Gazelle doesn't do that for us. (Not sure why not.) I'm hesitant to do this ugly thing because now we're trying to get `rules_go` to do something it doesn't want to do so our tests don't fail.

Bare minimum, you can add a new `go_library(name = "noop", importpath = "noop", visibility = ["//pkg:__subpackages__"])` in `build/bazelutil/BUILD.bazel` and reference that everywhere we want an empty `go_library`.

**Comment by rickystewart** [comment]
left additional commentary at https://github.com/bazelbuild/bazel-gazelle/issues/1078

File: pkg/kv/kvclient/rangefeed/BUILD.bazel
Comment (erikgrinaker) [inline pkg/kv/kvclient/rangefeed/BUILD.bazel:42]: I think we could end up needing multiple mock files -- in particular when using source mode, since I think `mockgen` then generates one output file per input file. With [Mockery](https://github.com/vektra/mockery), it places mocks in a `mocks/` subdirectory with `package mocks`. Perhaps a scheme like that could help with dependency handling as well?

File: pkg/kv/kvclient/kvcoord/BUILD.bazel
Comment (irfansharif) [reply pkg/kv/kvclient/kvcoord/BUILD.bazel]: Done. 

> The blessed way to do that would be to have two separate test targets, one for the external tests and one for the internal tests. Unfortunately Gazelle doesn't do that for us. (Not sure why not.)

Having separate test targets used to be a thing before, and was changed at some point: https://github.com/bazelbuild/bazel-gazelle/issues/102. Unsure why, brevity maybe? I assume `go_test` is internally capturing the fact that there are both internal and external tests specified in the same target, I think what I'm asking in  bazelbuild/bazel-gazelle#1078 then is for gazelle to construct the right dependencies by looking at the import path, rather than just the label. Curious to see what they think of this upstream.

File: pkg/kv/kvclient/rangefeed/BUILD.bazel
Comment (irfansharif) [reply pkg/kv/kvclient/rangefeed/BUILD.bazel:42]: The scheme here is generalizable to multiple mock files, we'd create a `gomock` target for each one and make sure they're listed as deps in the package's `:with-mocks` target. And yea, a separate mocks directory would help, and (test) packages that want to import mocks for another package, could import it directly. I think I'll avoid doing that for now and stick to the standard way of placing golang/mocks. We can always re-visit if this proves to be too cumbersome/decide to go with mockery.

Comment (erikgrinaker) [reply pkg/kv/kvclient/rangefeed/BUILD.bazel:42]: Yeah, I just find it a bit unfortunate that whenever someone wants to add a mock to a new package they'll not only have to muck around with the package build dependencies (which is fine enough), but also have to change the package target name, update dozens of unrelated build files, and fiddle with resolve overrides and such in the root build manifest.

It'd be nice if this was a bit less involved, and I'm thinking with separate mocks packages we might be able to get away with adding a new mock target that Gazelle could automatically depend on for any tests that need the mocks. I'm not entirely sure we can avoid circular dependencies in that case though, with internal tests that depend on the mocks.

Comment (irfansharif) [reply pkg/kv/kvclient/rangefeed/BUILD.bazel:42]: The two `gazelle:resolve` (in the top-level BUILD file + package's BUILD file for this annoying noop target) + `make bazel-generate` should do all of it for us. I'm also hoping neither will be necessary if the upstream issues are resolved/have easier work-arounds. Do you feel it's still too cumbersome? Or could be documented better? 

And yes, even with a separate mocks package, we'd need some bending over to work with internal tests. Curious to see if you have other ideas we could try.

Comment (erikgrinaker) [reply pkg/kv/kvclient/rangefeed/BUILD.bazel:42]: It's workable, and definitely an improvement over what we're currently doing, but a bit hacky. Don't have any better ideas off the bat though, and it's quite possible that this is as good as it's going to get right now. Might be worth looking around for some prior art.

**Comment by irfansharif** [comment]
Given the BUILD.bazel surface area of this PR, it keeps picking up merge conflicts against master. With a cursory browse around I couldn't find much in the way of alternatives, and I think it's a decent stopping point for now wrt mockgen + bazel (esp after filing the linked issues upstream). If y'all agree, can I get an LGTM?

**Comment by irfansharif** [comment]
(bump, I'm also happy to abandon this PR if we don't think it's sufficient.)

**Review by erikgrinaker — APPROVED** [review]
It's an improvement on the current state at least, so LGTM -- thank you for looking into this! Let's do another pass somewhere down the road.

**Comment by irfansharif** [comment]
TFTR!

bors r+

**Pull request merged**
